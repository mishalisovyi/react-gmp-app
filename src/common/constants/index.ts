export const Icon = {
  VERTICAL_ELLIPSIS: String.fromCharCode(8942),
  MULTIPLICATION_X: String.fromCharCode(10005),
  TELEPHONE_RECORDER: String.fromCharCode(8981),
};

export const API_URL = 'http://localhost:4000';

export const HTTP_RESPONSE_STATUS_NO_CONTENT = 204;
