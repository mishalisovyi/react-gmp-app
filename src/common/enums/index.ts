export { ButtonType } from './button-type.enum';
export { HttpMethod } from './http-method.enum';
