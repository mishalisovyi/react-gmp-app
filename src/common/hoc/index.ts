export { withPortalRendering } from './withPortalRendering/withPortalRendering';
export { ModalWindow } from './ModalWindow/ModalWindow';
