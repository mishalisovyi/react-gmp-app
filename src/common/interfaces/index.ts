export { EnumerableComponentProps } from './EnumerableComponentProps';
export { ModalWindowProps, ModalWindowActionButtonProps } from './ModalWindowProps';
export { SelectListData, SelectListValue } from './SelectListData';
