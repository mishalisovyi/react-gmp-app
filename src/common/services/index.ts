export {
  sendGetRequest,
  sendPostRequest,
  sendPutRequest,
  sendDeleteRequest,
} from './http.service';
