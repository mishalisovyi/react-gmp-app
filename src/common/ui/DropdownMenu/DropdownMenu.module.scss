@use '../../../colors';
@use '../../../mixins';

.DropdownMenu {
  position: absolute;
  display: flex;
  flex-direction: column;
  padding: 12px 0;
  background-color: colors.$young-night;
  box-shadow: 0px 0px 10px colors.$black-panther;
  transform: translateX(-100%);
  z-index: 1;

  &__close-button {
    align-self: end;
    margin: -10px 10px 4px 0;
  }

  &__item {
    @include mixins.typography-menu;
    padding: 8px 12px;
    min-width: 180px;
    color: colors.$white;
    cursor: pointer;

    &:hover {
      background-color: colors.$festive-fennec;
    }
  }
}