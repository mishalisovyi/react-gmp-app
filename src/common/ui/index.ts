export { Button } from './Button/Button';
export { Divider } from './Divider/Divider';
export { DropdownMenu } from './DropdownMenu/DropdownMenu';
export { FilterSelectList } from './FilterSelectList/FilterSelectList';
export { LabelPrimary } from './LabelPrimary/LabelPrimary';
export { LabelSecondary } from './LabelSecondary/LabelSecondary';
export { Logo } from './Logo/Logo';
export { PopupConfirmation } from './PopupConfirmation/PopupConfirmation';
export { PopupDialog } from './PopupDialog/PopupDialog';
export { PopupInfo } from './PopupInfo/PopupInfo';
export { SearchPanel } from './SearchPanel/SearchPanel';
export { Spinner } from './Spinner/Spinner';
export { Tab } from './TabBar/Tab/Tab';
export { TabBar } from './TabBar/TabBar';
export { TextField } from './TextField/TextField';
