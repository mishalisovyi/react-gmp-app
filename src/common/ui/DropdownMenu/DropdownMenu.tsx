import React, { useCallback, useState } from 'react';
import memoize from 'fast-memoize';

import { ButtonTransparent } from 'common/ui';
import styles from './DropdownMenu.module.scss';

interface DropdownMenuProps {
  children: React.ReactNode,
  items: string[],
  onOptionClicked: (option: string) => void;
}

export function DropdownMenu({ children, items, onOptionClicked }: DropdownMenuProps) {
  const [isOpened, setIsOpened] = useState(false);

  const handleMenuTargetClick = () => {
    setIsOpened(true);
  };

  const handleMenuOptionClick = useCallback(memoize((option: string) => {
    return () => {
      setIsOpened(false);
      onOptionClicked(option);
    };
  }), []);

  const handleCloseButtonClick = () => {
    setIsOpened(false);
  };

  return (
    <>
      {!isOpened && (
        <div onClick={handleMenuTargetClick}>{children}</div>
      )}

      {isOpened && (
        <div className={styles['DropdownMenu']}>
          <div className={styles['DropdownMenu__close-button']}>
            <ButtonTransparent text="&#10005;" onClick={handleCloseButtonClick} />
          </div>
          {items.map((item) => <div className={styles['DropdownMenu__item']} onClick={handleMenuOptionClick(item)} key={item}>{item}</div>)}
        </div>
      )}
    </>
  );
}
