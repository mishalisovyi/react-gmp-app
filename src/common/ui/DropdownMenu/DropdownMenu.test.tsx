import React from 'react';
import renderer from 'react-test-renderer';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { Icon } from 'common/constants';

import { DropdownMenu } from './DropdownMenu';

const menuItems = ['item 1', 'item 2'];
const menuTargetText = 'menu target';
const optionClickedHandler = jest.fn();

it('should render dropdown menu target element', () => {
  const tree = renderer.create(
    <DropdownMenu items={menuItems} onOptionClicked={optionClickedHandler}>
      <div>{menuTargetText}</div>
    </DropdownMenu>,
  );

  expect(tree.toJSON()).toMatchSnapshot();
});

it('should handle dropdown menu opening', async () => {
  render(
    <DropdownMenu items={menuItems} onOptionClicked={optionClickedHandler}>
      <div>{menuTargetText}</div>
    </DropdownMenu>,
  );

  await userEvent.click(screen.getByText(menuTargetText));

  expect(screen.getByText('item 1')).toBeInTheDocument();
});

it('should handle dropdown menu closing', async () => {
  render(
    <DropdownMenu items={menuItems} onOptionClicked={optionClickedHandler}>
      <div>{menuTargetText}</div>
    </DropdownMenu>,
  );

  await userEvent.click(screen.getByText(menuTargetText));
  await userEvent.click(screen.getByText(Icon.MULTIPLICATION_X));

  expect(screen.queryByText('item 1')).toBeNull();
});

it('should handle menu option click', async () => {
  render(
    <DropdownMenu items={menuItems} onOptionClicked={optionClickedHandler}>
      <div>{menuTargetText}</div>
    </DropdownMenu>,
  );

  await userEvent.click(screen.getByText(menuTargetText));
  await userEvent.click(screen.getByText(menuItems[0]));

  expect(optionClickedHandler).toHaveBeenCalledWith(menuItems[0]);
});
