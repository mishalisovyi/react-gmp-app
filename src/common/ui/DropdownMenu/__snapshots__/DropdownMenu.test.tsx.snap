// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render dropdown menu target element 1`] = `
<div
  onClick={[Function]}
>
  <div>
    menu target
  </div>
</div>
`;
